Specify checked, wrapping, overflowing, saturating operations on integer types

LevitatingLion opened the pull request:
And add the builtins `int.EuclideanDivision.div` as `Int::div_euclid` and `int.EuclideanDivision.mod` as `Int::rem_euclid`.

xldenis commented:
Awesome! Thanks for all these contributions. Could you add a test case verifying these specs work as intended?

xldenis commented on creusot-contracts/src/std/num.rs:
```diff
@@ -0,0 +1,85 @@
+
+/// Adds specifications for checked, wrapping, saturating, and overflowing versions of the given
+/// operation on the given type
+macro_rules! spec_op {
+    (
+        $type:ty,
+        $op:tt,
+        $checked:ident,
+        $wrapping:ident,
+        $saturating:ident,
+        $overflowing:ident $(,)?
+    ) => {
+        extern_spec! {
```
macros generating macros generating macros 😄

xldenis commented on creusot-contracts/src/std/num.rs:
```diff
@@ -0,0 +1,85 @@
+                fn $checked(self, rhs: $type) -> Option<$type>;
+
+                // Wrapping: performs the operation on `Int` and converts back to `$type`
+                #[allow(dead_code)]
```
You shouldn't need to `allow(dead_code)`.

xldenis commented on creusot-contracts/src/std/num.rs:
```diff
@@ -0,0 +1,85 @@
+                // Wrapping: performs the operation on `Int` and converts back to `$type`
+                #[allow(dead_code)]
+                #[ensures(
+                    @result == (@self $op @rhs).rem_euclid(2.pow(@$type::BITS)) + @$type::MIN
```
I wonder how usable this specification will be in practice. Perhaps there's a good way to lower this to SMT so that they will be able to reason (bit vectors?)

jhjourdan replied:
Probably we should have a post-condition saying its in bounds, and another saying they have the same value modulo  2^n, i.e., their difference is a multiple.

We should also have a post-condition saying that if self is in bounds, then the result is equal.

LevitatingLion commented:
I split division from the `spec_op!` macro since it has the additional precondition that the divisor is non-zero, and it only overflows in the single case of `MIN / -1` (which the prover is not able to deduce from the general specification).

I also added additional postconditions to `wrapping_*` and `overflowing_*`, asserting that the result is just the result of the operation if that is in range, and that it is offset by a multiple of the type's range otherwise (as suggested by @jhjourdan, and the prover is unable to verify all the testcases without these postconditions).

And I added some tests, which can be proved by CVC4 or Z3 except for the two `*_overflowing_mul`, which CVC4 is unable to prove (let me know if you have an idea how to fix that).

I agree that the `#[allow(dead_code)]` shouldn't be needed, but I do get warnings without them; probably caused by a specific interaction between `extern_spec!` and that the code inside is generated by a macro.

xldenis commented:
Thanks for addressing the comments and for the extensive test cases. I think this PR is ready to merge.

xldenis commented:
Looking at CI, I'll need to replay your proofs with a slightly older Z3, and longer term we need a CI image where we control the versions more tightly.

LevitatingLion commented:
I only tested the proofs with my local why3 installation, not in the CI image. I can do that tonight, maybe it fixes the version mismatch.

On August 2, 2022 1:53:10 PM GMT+02:00, Xavier Denis ***@***.***> wrote:
>Looking at CI, I'll need to replay your proofs with a slightly older Z3, and longer term we need a CI image where we control the versions more tightly.
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/xldenis/creusot/pull/505#issuecomment-1202385558
>You are receiving this because you authored the thread.
>
>Message ID: ***@***.***>

xldenis commented:
> I only tested the proofs with my local why3 installation, not in the CI image. I can do that tonight, maybe it fixes the version mismatch.

Ill just push a commit fixing them.

LevitatingLion commented:
Awesome!

xldenis commented:
> Awesome!

Thanks to you! I'm looking forward to getting your other PRs merged soon as well.
